### PR TITLE
New compiler directive to replace _compiler_version

### PR DESCRIPTION
**Pull request by hartbit**
This is implementation for the `Compiler Version Directive` proposal: https://github.com/apple/swift-evolution/pull/833

### CODE REVIEW

**Comment by tkremenek** [comment]
The implementation also needs to cover this [case](https://forums.swift.org/t/se-0212-compiler-version-directive/12267/7) that came up during the review evolution thread.

**Comment by hartbit** [comment]
@tkremenek Isn't that already the case in `Version::getEffectiveLanguageVersion()`? Can you point me as to how the implementation is incorrect?

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by tkremenek** [comment]
@hartbit you're absolutely right.  Sorry for the delay in responding.

@jrose-apple can you give this review with another pair of eyes?  Looks fairly straightforward.

File: lib/Parse/ParseIfConfig.cpp [hunk 1]
```diff
@@ -452,6 +453,13 @@ class EvaluateIfConfigCondition :
           Str, SourceLoc(), nullptr).getValue();
       auto thisVersion = Ctx.LangOpts.EffectiveLanguageVersion;
       return thisVersion >= Val;
+    } else if (KindName == "compiler") {
+      auto PUE = cast<PrefixUnaryExpr>(Arg);
+      auto Str = extractExprSource(Ctx.SourceMgr, PUE->getArg());
+      auto Val = version::Version::parseVersionString(
+          Str, SourceLoc(), nullptr).getValue();
+      auto thisVersion = version::Version::getCurrentLanguageVersion();
+      return thisVersion >= Val;
```
Comment (jrose-apple) [inline lib/Parse/ParseIfConfig.cpp]: It would be nice to share code with the `"swift"` case above!

**Review by jrose-apple — APPROVED** [review]
It looks good to me, with one potential cleanup comment.

**Comment by hartbit** [comment]
@swift-ci test

**Comment by hartbit** [comment]
@jrose-apple I addressed your cleanup comment. Thanks for the review!

**Review by jrose-apple — APPROVED** [review]

**Comment by tkremenek** [comment]
@swift-ci smoke test

**Comment by tkremenek** [comment]
@hartbit can you prepare another pull request against ‘swift-4.2-branch’?

**Pull request merged**